### ISSUE TITLE
Consolidate CI (yaml) files

### ISSUE THREAD

**Issue by Profpatsch** [tweag/rules_haskell#1100]
**Is your feature request related to a problem? Please describe.**

By now we have more than half a dozen different yaml files for CI purposes:

- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/.mergify.yml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/azure-pipelines.yml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/netlify.toml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/.bazelci/presubmit.yml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/.buildkite/pipeline.yml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/.circleci/config.yml
- https://github.com/tweag/rules_haskell/blob/5651181036ec08f77e91ebfe18138b3a9e0ae7f4/.github/setting.yml

None of it is particularly descriptive and stuff is duplicated over multiple files.

**Describe the solution you'd like**

- Deploy a [dhall](https://dhall-lang.org) configuration with a clear structure and meta-fields describing what each configuration does.
  - Duplicated content can be factored out into simple functions
- Use a small script to generate all the config files from the dhall configuration
  - Each generated file gets a header which points towards the original dhall description
- Add a CI lint rule that checks whether all configurations have been updated

It won’t reduce the clutter of files (we’ll have to check in the generated files anyway), but we get a reasonable overview of what CI we are using, who can be asked if there are questions, and why we run every single CI.

**Describe alternatives you've considered**

- Leave it like it is
  - Contributors have a very hard time grokking all the different setups
  - The bus factor is very low, it’s not documented who has ownership over which file
  - There is no meta-information about what each file is for

- Leave it like it is, symlink each CI file from a central directory
  - Viable alternative
  - Might give us problems with azure (because Windows doesn’t like symlinks)
  - No deduplication
  - yaml is easy to get wrong, especially concerning mutliline strings (and strange value type conversions), so time is wasted on debugging CI problems stemming from yaml weirdness  
  - Can be combined with the dhall approach, generating files in a separate directory and symlinking them to their positions afterwards

- Use nix to generate the files
  On the lorri repo [we do something similar](https://github.com/target/lorri/blob/master/.travis.yml.nix), but we use the nix evaluator instead of dhall
  - Nix is untyped, so it’s harder to understand what functions do
  - It’s a side-effect, not what Nix is originally intended to do
  - `dhall` is a self-contained, static binary while nix has to be installed permanently on the system

**Comment by mboes** [comment]
I'm loath to introduce extra indirections to generate output that ultimately needs to be checked-in, because various tools will be looking for these YAML files in various hardcoded places that we can't control.

I'm not sure what "problem" you have identified in the problem description, other than some things being duplicated. For that matter, I'm finding it hard to identify much duplication. Do you have examples? There's duplication of `build_tag_filters` in a couple of places. This could be avoided by defining a `test_suite` in Bazel BUILD files. Further command-line arguments, like platforms, will in the near future also be avoidable by pushing still more in BUILD files.

So to summarize, I'm skeptical that the indirection will save us any lines of code, there isn't much to abstract, and the little there is to abstract should be pushed into BUILD files, not yet more external configuration.

**Comment by aherrmann** [comment]
This list is now down to the following files

* CI
  * .github/workflows/workflow.yaml 
  * netlify.toml
  * .bazelci/presubmit.yml
* Other
  * .mergify.yml
  * .github/settings.yml

The first three define actual CI pipelines, the latter two define other CI related configuration.

The build configuration itself has also largely been concentrated in `.bazelrc` by now.

There is relatively little overlap left between these files. What comes to mind are

* The names of configurations (Bazel `--config` flags)
* The name of the CI pipelines for the mergify check
* The name of the mergify label

I don't think the cost of introducing a configuration generator outweights the benefit at this point.
